examples/optim.jl

```diff
@@ -0,0 +1,219 @@
+# # Optimizing agent-based models
+
+# Agent-based models (ABMs) are computationally more expensive than analytical models, and can have many parameters. Sometimes we need to fine-tune a model's parameters to a specific outcome. Brute-force algorithms can take too long for testing each parameter setting. Even if it was feasbile to run the model for every parameter setting, it would not be enough because ABMs are stochastic and the effect of a parameter setting should be derived from running the model several times and taking its average behavior.
+
+# Here we show how to use the evolutionary algorithms in [BlackBoxOptim.jl](https://github.com/robertfeldt/BlackBoxOptim.jl) with Agents.jl, to optimize the parameters of an epidemiological model (SIR). We explain this model in detail in [SIR model for the spread of COVID-19](@ref). For brevity here, we just import it.
+
+cd(@__DIR__) #src
+cd("../../../examples/") #src
+include("siroptim.jl") ## From the examples directory
+
+# Now we need to define a cost function. The cost function takes as agruments the model parameters that we want to tune, here migration rate, death rate, transmission rate when an infected person has been (not) detected (`β_det`, `β_und`), infection period, reinfection probability, and time until the infection is detected. The function returns one or more numbers as the objective to be minimized. Here, we try to minimize the number of infected people after 50 days.
+
+using BlackBoxOptim
+import Statistics: mean
+
+function cost(x)
+    migration_rate,
+    death_rate,
+    β_det,
+    β_und,
+    infection_period,
+    reinfection_probability,
+    detection_time = x
+    C = 3
+    params = create_params(
+        C = C,
+        Ns = [500 for i in 1:C],
+        β_det = [β_det for i in 1:C],
+        migration_rate = migration_rate,
+        infection_period = infection_period,
+        reinfection_probability = reinfection_probability,
+        detection_time = detection_time,
+        death_rate = death_rate,
+        Is = ones(Int, C),
+        β_und = [β_und for i in 1:C],
+    )
+
+    model = model_initiation(; params...)
+
+    infected_fraction(model) =
+        count(a.status == :I for a in values(model.agents)) / nagents(model)
+    _, data = run!(
+        model,
```

Comment:
there seems to be some missalignment here

Reply:
I am not sure why we can't directly pass `x` to model creation...? It seems we do a lot of unecessary `reinfection_probability, ... = x` and then again `reinfection_probability = reinfection_probability` why is there so much duplication?

Reply:
Looking into better refactors of this and the other exposed functions from siroptim. If this needs to happen due to the optimisation steps, it should certainly be documented further.